server/stream.go

```diff
@@ -4099,41 +4098,86 @@ func (mset *stream) processDirectGetLastBySubjectRequest(_ *subscription, c *cli
 // Do actual work on a direct msg request.
 // This could be called in a Go routine if we are inline for a non-client connection.
 func (mset *stream) getDirectRequest(req *JSApiMsgGetRequest, reply string) {
-	var svp StoreMsg
-	var sm *StoreMsg
-	var err error
-
 	mset.mu.RLock()
-	store, name := mset.store, mset.cfg.Name
+	store, name, s := mset.store, mset.cfg.Name, mset.srv
 	mset.mu.RUnlock()
 
-	if req.Seq > 0 && req.NextFor == _EMPTY_ {
-		sm, err = store.LoadMsg(req.Seq, &svp)
-	} else if req.NextFor != _EMPTY_ {
-		sm, _, err = store.LoadNextMsg(req.NextFor, subjectHasWildcard(req.NextFor), req.Seq, &svp)
-	} else {
-		sm, err = store.LoadLastMsg(req.LastFor, &svp)
-	}
-	if err != nil {
-		hdr := []byte("NATS/1.0 404 Message Not Found\r\n\r\n")
-		mset.outq.send(newJSPubMsg(reply, _EMPTY_, _EMPTY_, hdr, nil, nil, 0))
-		return
-	}
+	seq := req.Seq
+	wc := subjectHasWildcard(req.NextFor)
+	batch := req.Batch
+	if batch == 0 {
+		batch = 1
+	}
+	// Grab MaxBytes
+	mb := req.MaxBytes
+	if mb == 0 && s != nil {
+		// Fill in with the server's MaxPending.
+		mb = int(s.opts.MaxPending)
```

Comment:
Should we be capping at `MaxPending` even if `MaxBytes` is specified?

Reply:
I don't think so. My thinking was if you know you are on a 100Gbit line and can easily sustain, we should not stop you. If you specify it and you become a slow consumer, pretty clear signal you need to adjust it.